lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3793,11 +4300,38 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
     }
     auto cleanup = makeCleanup(adjBuf,
                                [](SILBuilder &b, SILLocation l, SILValue v) {
-      b.createReleaseValueAddr(l, v, b.getDefaultAtomicity());
+      if (v->getType().isLoadable(b.getModule()))
+        b.createReleaseValueAddr(l, v, b.getDefaultAtomicity());
+      else
+        b.createDestroyAddr(l, v);
     });
     adjBuf.setCleanup(cleanup);
   }
 
+  // Handle `copy_addr` instruction.
+  //   Original: copy_addr x to y
+  //    Adjoint: adj[x] += adj[y]; adj[y] = 0
+  void visitCopyAddrInst(CopyAddrInst *cai) {
+    auto adjDest = getAdjointBuffer(cai->getDest());
+    // Disable the buffer's top-level cleanup (which is supposed to operate on
+    // the buffer), create a cleanup for the value that carrys all child
+    // cleanups.
+    auto valueCleanup = makeCleanup(adjDest,
+        [](SILBuilder &b, SILLocation l, SILValue v) {
+          b.createReleaseValue(l, v, b.getDefaultAtomicity());
+        }, adjDest.getCleanup()
+            ? adjDest.getCleanup()->getChildren() : ArrayRef<Cleanup *>());
+    adjDest.setCleanup(valueCleanup);
+    auto *readAccess = builder.createBeginAccess(
+        cai->getLoc(), adjDest, SILAccessKind::Read,
+        SILAccessEnforcement::Static, /*noNestedConflict*/ true,
+        /*fromBuiltin*/ false);
+    addToAdjointBuffer(cai->getSrc(), readAccess);
+    builder.createEndAccess(cai->getLoc(), readAccess, /*aborted*/ false);
+    auto destType = remapType(adjDest.getType());
+    emitZeroIndirect(destType.getASTType(), adjDest, cai->getLoc());
```

Comment:
`emitZeroIndirect` takes an access. You need `begin_access [modify]` and `end_access`.

Reply:
Zero needs a new cleanup. You should do what `visitStoreInst` is doing, including matching access kinds.

Reply:
Done in https://github.com/apple/swift/pull/22544/commits/ad7436a8e38562bbf4aa6cf6ebe6918c829c5331.